go/signedexchange/signedexchange_test.go

```diff
@@ -0,0 +1,181 @@
+IcMi5IPGkhHqpvxKvnHnHnYZJldNfTvjQykcAXmUiqkFCE41XYBPSj0cx472hiaE
+hPGHSUdaaaRBbsbVOy/aZSRFBIA8ngxyrW6B94Q/uLVZBn6axqoj8xT1YFVBgH5G
+/lVxfkpjUD2im9r3w+7ofSmMKa6CyJ/bBdRf8p0ACyzDbkfyXjwUxSj/ZFrpLg66
+amEXgauqxKEAhF8MP8oKir9aEwl7EaYFIRFpzQ6LT6edD5vcieov6hDi1f8xxdty
+5lL4HkECgYEA01+pVvn2VqANu9tgpcX3srY6QKnqViBSXr6GX+XpcCJlxR2S4FVD
+gdEwMHJK9137krvzIek57BFQXd4bTpeUW3Da8rX73tUnqKrQ5pmEqpghRyCqo0kT
+V1ObepNUcQVmK6VnqIuckHNV7sjYnSCgY4P4WiPBRJCG3jTI2LUpo/UCgYEAwrV9
+MtwsV9HlVHNrd8hqqaXnDvY1InFCfFxyR0m5KMTiwvcswBbwpTYtKZXWnz2HRVbO
+aMmh2RQKk9Swpwb/q2TjVnPPUqH14++OwyR0k/0L4KBZMY736GqyWnfod6G5KQD2
+f5MtwRFCYoJ6Tts4KtMzxxaV4TeRQA0EES7rK/0CgYBVztbi7TSYs/7/TS6t/XDx
+xtJdH912u0ZVGglY8u/SStR/seLHWTW/hJmIgU13oFqZld083f5anCjBAoKZZCWg
+/W6U61XlfyjLaxTFGHtn+bxAsL007lyArftHRnoYK7XvcAVlwc98QKYY+sYc+3rB
+C3kNtsglunpVyJ3kg5705QJ/cVMwi2maZYLE92I2KoF7k0H8ObkTM/i3uaoU2WkP
+W6s8UD2MzkCLz5y4rHuJbyVglfrwKA0zJiWEAobISm7IX/lYV/kPsgiSFRhY/zs4
+numpABRT1YRgxeVT6VPg+cAnBLaKwbXn63cgLDXE+iCdkE9c04NRuMOexqjMtTOZ
+rQKBgDSCTKwnbJUqN94WdBYjinFN/bR6E0wW640jkB/3e8Y4a+W4OVHWlxoEu4Tm
+s5B6gZsV/ojttR+aaeRknfrhQwEIA/k2r2oZE9yp8djzyiiqGswgw8yO0WSJztbx
+GRqzPwjon7ESIVpKLrVuh5qlMhUkOFUeF9wvViWX4qnV5Fvg
+-----END RSA PRIVATE KEY-----
+`
+)
+
+type zeroReader struct{}
+
+func (zeroReader) Read(b []byte) (int, error) {
+	for i := range b {
+		b[i] = 0
+	}
+	return len(b), nil
+}
+
+func mustReadFile(path string) []byte {
+	b, err := ioutil.ReadFile(path)
+	if err != nil {
+		panic(err)
+	}
+	return b
+}
+
+func TestSignedExchange(t *testing.T) {
+	u, _ := url.Parse("https://example.com/")
+	header := http.Header{}
+	header.Add("Content-Type", "text/html; charset=utf-8")
+	e, err := signedexchange.NewExchange(u, 200, header, []byte(payload), 16)
+	if err != nil {
+		t.Fatal(err)
+	}
+	e.AddSignedHeadersHeader("Content-Type", "Content-Encoding", "MI")
+
+	now := time.Date(2018, 1, 31, 17, 13, 20, 0, time.UTC)
+	certs, err := signedexchange.ParseCertificates([]byte(pemCerts))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	derPrivateKey, _ := pem.Decode([]byte(pemPrivateKey))
+	privKey, err := signedexchange.ParsePrivateKey(derPrivateKey.Bytes)
+	if err != nil {
+		t.Fatal(err)
+	}
+	certUrl, _ := url.Parse("https://example.com/cert.msg")
+	validityUrl, _ := url.Parse("https://example.com/resource.validity")
+	s := &signedexchange.Signer{
+		Date:        now,
+		Expires:     now.Add(1 * time.Hour),
+		Certs:       certs,
+		CertUrl:     certUrl,
+		ValidityUrl: validityUrl,
+		PrivKey:     privKey,
+		Rand:        zeroReader{},
+	}
+	if err := e.AddSignatureHeader(s); err != nil {
+		t.Fatal(err)
+	}
+
+	var buf bytes.Buffer
+	signedexchange.WriteExchangeFile(&buf, e)
+
+	var decoded interface{}
+	handle := &codec.CborHandle{}
+	if err := codec.NewDecoder(&buf, handle).Decode(&decoded); err != nil {
+		t.Fatal(err)
+	}
+	got := strings.TrimSpace(fmt.Sprintf("%q", decoded))
+	want := strings.TrimSpace(string(mustReadFile("test-signedexchange-expected.txt")))
```

Comment:
Found that comparing `%q` strings is fragile since the order of map keys is not deterministic. I'll fix this later.

Reply:
Done.